examples/src/main/java/com/mapbox/navigation/examples/core/FreeDriveNavigationActivity.kt

```diff
@@ -0,0 +1,237 @@
+            } else {
+                startNavigation.visibility = View.GONE
+            }
+        }
+
+        override fun onRoutesRequestFailure(throwable: Throwable, routeOptions: RouteOptions) {
+            Timber.e("route request failure %s", throwable.toString())
+        }
+
+        override fun onRoutesRequestCanceled(routeOptions: RouteOptions) {
+            Timber.d("route request canceled")
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    fun initListeners() {
+        startNavigation.setOnClickListener {
+            navigationMapboxMap?.enterNavigationMode()
+            navigationMapboxMap?.addProgressChangeListener(mapboxNavigation!!)
+            if (mapboxNavigation?.getRoutes()?.isNotEmpty() == true) {
+                navigationMapboxMap?.startCamera(mapboxNavigation?.getRoutes()!![0])
+            }
+            mapboxNavigation?.startTripSession()
+            startNavigation.visibility = View.GONE
+        }
+    }
+
+    override fun onStart() {
+        super.onStart()
+        mapView.onStart()
+        mapboxNavigation?.registerLocationObserver(locationObserver)
+    }
+
+    public override fun onResume() {
+        super.onResume()
+        mapView.onResume()
+    }
+
+    public override fun onPause() {
+        super.onPause()
+        mapView.onPause()
+    }
+
+    override fun onStop() {
+        super.onStop()
+        mapboxNavigation?.unregisterLocationObserver(locationObserver)
+        mapboxNavigation?.unregisterRouteProgressObserver(routeProgressObserver)
+        stopLocationUpdates()
+        mapView.onStop()
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        mapboxNavigation?.stopTripSession()
+        mapboxNavigation?.onDestroy()
+        mapView.onDestroy()
+    }
+
+    override fun onLowMemory() {
+        super.onLowMemory()
+        mapView.onLowMemory()
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        navigationMapboxMap?.saveStateWith(MAP_INSTANCE_STATE_KEY, outState)
+        mapView.onSaveInstanceState(outState)
+    }
+
+    override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
+        super.onRestoreInstanceState(savedInstanceState)
+        mapInstanceState = savedInstanceState?.getParcelable(MAP_INSTANCE_STATE_KEY)
+    }
+
+    private val locationListenerCallback: LocationEngineCallback<LocationEngineResult> =
+            object : LocationEngineCallback<LocationEngineResult> {
+                override fun onSuccess(result: LocationEngineResult) {
+                    // todo
+                }
+
+                override fun onFailure(exception: Exception) {
+                    Timber.i(exception)
+                }
+            }
+
+    private fun stopLocationUpdates() {
+        mapboxNavigation?.locationEngine?.removeLocationUpdates(locationListenerCallback)
+    }
+
+    private val locationObserver = object : LocationObserver {
```

Comment:
Do we as a client need to push Nav Core location updates manually? Couldn't `NavigationMapboxMap` subscribe to Nav Core's location observer when we pass it in `addProgressChangeListener` (can be renamed if needed)?

Reply:
I had the same thought but I didn't have enough context to know why this hadn't already been implemented. I"ll do some experimenting.